wgc/wgc_keccak.py

```diff
@@ -0,0 +1,342 @@
+    padlen = align_bytes - used_bytes
+    if padlen == 0:
+        padlen = align_bytes
+    # note: padding done in 'internal bit ordering', wherein LSB is leftmost
+    if padlen == 1:
+        return [0x81]
+    else:
+        return [0x01] + ([0x00] * (padlen - 2)) + [0x80]
+
+def keccak_f(state):
+    """
+    This is Keccak-f permutation.  It operates on and
+    mutates the passed-in KeccakState.  It returns nothing.
+    """
+
+    def round(a, rc):
+        w, h = state.w, state.h
+        rangew, rangeh = state.rangeW, state.rangeH
+        lanew = state.lanew
+        zero = state.zero
+
+        # theta
+        c = [reduce(xor, a[x]) for x in rangew]
+        d = [0] * w
+        for x in rangew:
+            d[x] = c[(x - 1) % w] ^ rol(c[(x + 1) % w], 1, lanew)
+            for y in rangeh:
+                a[x][y] ^= d[x]
+
+        # rho and pi
+        b = zero()
+        for x in rangew:
+            for y in rangeh:
+                b[y % w][(2 * x + 3 * y) % h] = rol(a[x][y], RotationConstants[y][x], lanew)
+
+        # chi
+        for x in rangew:
+            for y in rangeh:
+                a[x][y] = b[x][y] ^ ((~ b[(x + 1) % w][y]) & b[(x + 2) % w][y])
+
+        # iota
+        a[0][0] ^= rc
+
+    nr = 12 + 2 * int(log(state.lanew, 2))
+
+    for ir in range(nr):
+        round(state.s, RoundConstants[ir])
+
+
+class KeccakState(object):
+    """
+    A keccak state container.
+
+    The state is stored as a 5x5 table of integers.
+    """
+    w = 5
+    h = 5
+
+    rangeW = range(w)
+    rangeH = range(h)
+
+    @staticmethod
+    def zero():
+        """
+        Returns an zero state table.
+        """
+        return [[0] * KeccakState.w for _ in KeccakState.rangeH]
+
+    @staticmethod
+    def format(st):
+        """
+        Formats the given state as hex, in natural byte order.
+        """
+        rows = []
+
+        def fmt(stx):
+            return '%016x' % stx
+
+        for y in KeccakState.rangeH:
+            row = []
+            for x in KeccakState.rangeW:
+                row.append(fmt(st[x][y]))
+            rows.append(' '.join(row))
+        return '\n'.join(rows)
+
+    @staticmethod
+    def lane2bytes(s, w):
+        """
+        Converts the lane s to a sequence of byte values,
+        assuming a lane is w bits.
+        """
+        o = []
+        for b in range(0, w, 8):
+            o.append((s >> b) & 0xff)
+        return o
+
+    @staticmethod
+    def bytes2lane(bb):
+        """
+        Converts a sequence of byte values to a lane.
+        """
+        r = 0
+        for b in reversed(bb):
+            r = r << 8 | b
+        return r
+
+    @staticmethod
+    def bytes2str(bl: list) -> bytes:
+        """
+        Converts a sequence of byte values to a string.
+        """
+        return bytes(bl)
+
+    @staticmethod
+    def str2bytes(bs: bytes) -> list:
+        """
+        Converts a string to a sequence of byte values.
+        """
+        return list(bs)
```

Comment:
Probably something wrong with docstrings

Reply:
Nope, it is correct in this case. `bytes` represents here ASCII-encoded string

Reply:
simplified in 31b6f6954b307505e25dca6a62030e14878adcfc